File: src/engine/Container.ts
import { StringValue } from "./Value";
import { throwNullException } from "./NullException";
import { StringBuilder } from "./StringBuilder";
import { INamedContent } from "./INamedContent";
import { InkObject } from "./Object";
import { SearchResult } from "./SearchResult";
import { Path } from "./Path";
import { Debug } from "./Debug";
import { tryGetValueFromMap } from "./TryGetResult";
import { asINamedContentOrNull, asOrNull, asOrThrows } from "./TypeAssertion";

export class Container extends InkObject implements INamedContent {
  public name: string | null = null;

  public _content: InkObject[] = [];
  public namedContent: Map<string, INamedContent> = new Map();

  public visitsShouldBeCounted: boolean = false;
  public turnIndexShouldBeCounted: boolean = false;
  public countingAtStartOnly: boolean = false;

  public _pathToFirstLeafContent: Path | null = null;

  get hasValidName() {
    return this.name != null && this.name.length > 0;
  }
  get content() {
    return this._content;
  }
  set content(value: InkObject[]) {
    this.AddContent(value);
  }
  get namedOnlyContent() {
    let namedOnlyContentDict: Map<string, InkObject> | null = new Map();

    for (let [key, value] of this.namedContent) {
      let inkObject = asOrThrows(value, InkObject);
      namedOnlyContentDict.set(key, inkObject);
    }

    for (let c of this.content) {
      let named = asINamedContentOrNull(c);
      if (named != null && named.hasValidName) {
        namedOnlyContentDict.delete(named.name!);
      }
    }

    if (namedOnlyContentDict.size == 0) namedOnlyContentDict = null;

    return namedOnlyContentDict;
  }
  set namedOnlyContent(value: Map<string, InkObject> | null) {
    let existingNamedOnly = this.namedOnlyContent;
    if (existingNamedOnly != null) {
      for (let [key] of existingNamedOnly) {
        this.namedContent.delete(key);
      }
    }

    if (value == null) return;

    for (let [, val] of value) {
      let named = asINamedContentOrNull(val);
      if (named != null) this.AddToNamedContentOnly(named);
    }
  }
  get countFlags(): number {
    let flags: Container.CountFlags = 0;
    if (this.visitsShouldBeCounted) flags |= Container.CountFlags.Visits;
    if (this.turnIndexShouldBeCounted) flags |= Container.CountFlags.Turns;
    if (this.countingAtStartOnly) flags |= Container.CountFlags.CountStartOnly;

    if (flags == Container.CountFlags.CountStartOnly) {
      flags = 0;
    }

    return flags;
  }
  set countFlags(value: number) {
    let flag: Container.CountFlags = value;
    if ((flag & Container.CountFlags.Visits) > 0)
      this.visitsShouldBeCounted = true;
    if ((flag & Container.CountFlags.Turns) > 0)
      this.turnIndexShouldBeCounted = true;
    if ((flag & Container.CountFlags.CountStartOnly) > 0)
      this.countingAtStartOnly = true;
  }
  get pathToFirstLeafContent() {
    if (this._pathToFirstLeafContent == null)
      this._pathToFirstLeafContent = this.path.PathByAppendingPath(
        this.internalPathToFirstLeafContent
      );

    return this._pathToFirstLeafContent;
  }
  get internalPathToFirstLeafContent() {
    let components: Path.Component[] = [];
    let container: Container = this;
    while (container instanceof Container) {
      if (container.content.length > 0) {
        components.push(new Path.Component(0));
        container = container.content[0] as Container;
      }
    }
    return new Path(components);
  }

  public AddContent(contentObjOrList: InkObject | InkObject[]) {
    if (contentObjOrList instanceof Array) {
      let contentList = contentObjOrList as InkObject[];

      for (let c of contentList) {
        this.AddContent(c);
      }
    } else {
      let contentObj = contentObjOrList as InkObject;

      this._content.push(contentObj);

      if (contentObj.parent) {
        throw new Error("content is already in " + contentObj.parent);
      }

      contentObj.parent = this;

      this.TryAddNamedContent(contentObj);
    }
  }
  public TryAddNamedContent(contentObj: InkObject) {
    let namedContentObj = asINamedContentOrNull(contentObj);
    if (namedContentObj != null && namedContentObj.hasValidName) {
      this.AddToNamedContentOnly(namedContentObj);
    }
  }
  public AddToNamedContentOnly(namedContentObj: INamedContent) {
    Debug.AssertType(
      namedContentObj,
      InkObject,
      "Can only add Runtime.Objects to a Runtime.Container"
    );
    let runtimeObj = asOrThrows(namedContentObj, InkObject);
    runtimeObj.parent = this;

    if (namedContentObj.name === null)
      return throwNullException("namedContentObj.name");
    this.namedContent.set(namedContentObj.name!, namedContentObj);
  }
  public ContentAtPath(
    path: Path,
    partialPathStart: number = 0,
    partialPathLength: number = -1
  ) {
    if (partialPathLength == -1) partialPathLength = path.length;

    let result = new SearchResult();
    result.approximate = false;

    let currentContainer: Container | null = this;
    let currentObj: InkObject = this;

    for (let i = partialPathStart; i < partialPathLength; ++i) {
      let comp = path.GetComponent(i);
      if (currentContainer == null) {
        result.approximate = true;
        break;
      }

      let foundObj: InkObject | null =
        currentContainer.ContentWithPathComponent(comp);

      // Couldn't resolve entire path?
      if (foundObj == null) {
        result.approximate = true;
        break;
      }

      // Are we about to loop into another container?
      // Is the object a container as expected? It might
      // no longer be if the content has shuffled around, so what
      // was originally a container no longer is.
      const nextContainer: Container | null = asOrNull(foundObj, Container);
      if (i < partialPathLength - 1 && nextContainer == null) {
        result.approximate = true;
        break;
      }

      currentObj = foundObj;
      currentContainer = nextContainer;
    }

    result.obj = currentObj;

    return result;
  }
  public InsertContent(contentObj: InkObject, index: number) {
    this.content.splice(index, 0, contentObj);

    if (contentObj.parent) {
      throw new Error("content is already in " + contentObj.parent);
    }

    contentObj.parent = this;

    this.TryAddNamedContent(contentObj);
  }
  public AddContentsOfContainer(otherContainer: Container) {
    this.content.push(...otherContainer.content);

    for (let obj of otherContainer.content) {
      obj.parent = this;
      this.TryAddNamedContent(obj);
    }
  }
  public ContentWithPathComponent(component: Path.Component): InkObject | null {
    if (component.isIndex) {
      if (component.index >= 0 && component.index < this.content.length) {
        return this.content[component.index];
      } else {
        return null;
      }
    } else if (component.isParent) {
      return this.parent;
    } else {
      if (component.name === null) {
        return throwNullException("component.name");
      }
      let foundContent = tryGetValueFromMap(
        this.namedContent,
        component.name,
        null
      );
      if (foundContent.exists) {
        return asOrThrows(foundContent.result, InkObject);
      } else {
        return null;
      }
    }
  }
  public BuildStringOfHierarchy(): string;
  public BuildStringOfHierarchy(
    sb: StringBuilder,
    indentation: number,
    pointedObj: InkObject | null
  ): string;
  public BuildStringOfHierarchy() {
    let sb: StringBuilder;
    if (arguments.length == 0) {
      sb = new StringBuilder();
      this.BuildStringOfHierarchy(sb, 0, null);
      return sb.toString();
    }

    sb = arguments[0] as StringBuilder;
    let indentation = arguments[1] as number;
    let pointedObj = arguments[2] as InkObject | null;

    function appendIndentation() {
      const spacesPerIndent = 4; // Truly const in the original code
      for (let i = 0; i < spacesPerIndent * indentation; ++i) {
        sb.Append(" ");
      }
    }

    appendIndentation();
    sb.Append("[");

    if (this.hasValidName) {
      sb.AppendFormat(" ({0})", this.name);
    }

    if (this == pointedObj) {
      sb.Append("  <---");
    }

    sb.AppendLine();

    indentation++;

    for (let i = 0; i < this.content.length; ++i) {
      let obj = this.content[i];

      if (obj instanceof Container) {
        let container = obj as Container;

        container.BuildStringOfHierarchy(sb, indentation, pointedObj);
      } else {
        appendIndentation();
        if (obj instanceof StringValue) {
          sb.Append('"');
          sb.Append(obj.toString().replace("\n", "\\n"));
          sb.Append('"');
        } else {
          sb.Append(obj.toString());
        }
      }

      if (i != this.content.length - 1) {
        sb.Append(",");
      }

      if (!(obj instanceof Container) && obj == pointedObj) {
        sb.Append("  <---");
      }

      sb.AppendLine();
    }

    let onlyNamed: Map<string, INamedContent> = new Map();

    for (let [key, value] of this.namedContent) {
      if (this.content.indexOf(asOrThrows(value, InkObject)) >= 0) {
        continue;
      } else {
        onlyNamed.set(key, value);
      }
    }

    if (onlyNamed.size > 0) {
      appendIndentation();
      sb.AppendLine("-- named: --");

      for (let [, value] of onlyNamed) {
        Debug.AssertType(
          value,
          Container,
          "Can only print out named Containers"
        );
        let container = value as Container;
        container.BuildStringOfHierarchy(sb, indentation, pointedObj);
        sb.AppendLine();
      }
    }

    indentation--;

    appendIndentation();
    sb.Append("]");
  }
}

export namespace Container {
  export enum CountFlags {
    Start = 0,
    Visits = 1,
    Turns = 2,
    CountStartOnly = 4,
  }
}

File: ink_dart/lib/src/engine/container.dart
import 'ink_object.dart';
import 'path.dart';
import 'value.dart';

/// Flags for visit counting behavior.
class CountFlags {
  static const int visits = 1;
  static const int turns = 2;
  static const int countStartOnly = 4;
}

/// Container that holds story content in a hierarchy.
///
/// Containers can hold:
/// - Other containers (knots, stitches)
/// - Text content
/// - Control commands
/// - Diverts, choices, etc.
class Container extends InkObject with INamedContent {
  @override
  String? name;

  final List<InkObject> _content = [];
  final Map<String, INamedContent> namedContent = {};

  bool visitsShouldBeCounted = false;
  bool turnIndexShouldBeCounted = false;
  bool countingAtStartOnly = false;

  Path? _pathToFirstLeafContent;

  @override
  bool get hasValidName => name != null && name!.isNotEmpty;

  /// The content objects in this container.
  List<InkObject> get content => _content;

  /// Set content, adding all items.
  set content(List<InkObject> value) {
    addContentList(value);
  }

  /// Get named-only content (not in content list).
  Map<String, InkObject>? get namedOnlyContent {
    final namedOnly = <String, InkObject>{};

    for (final entry in namedContent.entries) {
      if (entry.value is InkObject) {
        namedOnly[entry.key] = entry.value as InkObject;
      }
    }

    for (final c in content) {
      if (c is INamedContent) {
        final named = c as INamedContent;
        if (named.hasValidName && named.name != null) {
          namedOnly.remove(named.name);
        }
      }
    }

    return namedOnly.isEmpty ? null : namedOnly;
  }

  /// Set named-only content.
  set namedOnlyContent(Map<String, InkObject>? value) {
    final existing = namedOnlyContent;
    if (existing != null) {
      for (final key in existing.keys) {
        namedContent.remove(key);
      }
    }

    if (value == null) return;

    for (final val in value.values) {
      if (val is INamedContent) {
        addToNamedContentOnly(val as INamedContent);
      }
    }
  }

  /// Packed count flags for serialization.
  int get countFlags {
    int flags = 0;
    if (visitsShouldBeCounted) flags |= CountFlags.visits;
    if (turnIndexShouldBeCounted) flags |= CountFlags.turns;
    if (countingAtStartOnly) flags |= CountFlags.countStartOnly;

    // CountStartOnly is meaningless without Visits or Turns
    if (flags == CountFlags.countStartOnly) {
      flags = 0;
    }

    return flags;
  }

  /// Unpack count flags from serialization.
  set countFlags(int value) {
    if ((value & CountFlags.visits) > 0) visitsShouldBeCounted = true;
    if ((value & CountFlags.turns) > 0) turnIndexShouldBeCounted = true;
    if ((value & CountFlags.countStartOnly) > 0) countingAtStartOnly = true;
  }

  /// Path to the first leaf (non-container) content.
  Path get pathToFirstLeafContent {
    _pathToFirstLeafContent ??= path.pathByAppendingPath(internalPathToFirstLeafContent);
    return _pathToFirstLeafContent!;
  }

  /// Internal relative path to first leaf content.
  Path get internalPathToFirstLeafContent {
    final components = <PathComponent>[];
    InkObject current = this;

    while (current is Container) {
      final container = current;
      if (container.content.isNotEmpty) {
        components.add(PathComponent.fromIndex(0));
        current = container.content[0];
      } else {
        break;
      }
    }

    return Path.fromComponents(components);
  }

  /// Add a single content object.
  void addContent(InkObject contentObj) {
    _content.add(contentObj);

    if (contentObj.parent != null) {
      throw StateError('content is already in ${contentObj.parent}');
    }

    contentObj.parent = this;
    tryAddNamedContent(contentObj);
  }

  /// Add multiple content objects.
  void addContentList(List<InkObject> contentList) {
    for (final c in contentList) {
      addContent(c);
    }
  }

  /// Try to add an object to named content if it has a valid name.
  void tryAddNamedContent(InkObject contentObj) {
    if (contentObj is INamedContent) {
      final named = contentObj as INamedContent;
      if (named.hasValidName) {
        addToNamedContentOnly(named);
      }
    }
  }

  /// Add to named content only (not to content list).
  void addToNamedContentOnly(INamedContent namedContentObj) {
    if (namedContentObj is! InkObject) {
      throw ArgumentError('Can only add InkObjects to a Container');
    }

    final runtimeObj = namedContentObj as InkObject;
    runtimeObj.parent = this;

    if (namedContentObj.name == null) {
      throw StateError('namedContentObj.name is null');
    }

    namedContent[namedContentObj.name!] = namedContentObj;
  }

  /// Find content at a path.
  SearchResult contentAtPath(
    Path path, {
    int partialPathStart = 0,
    int partialPathLength = -1,
  }) {
    if (partialPathLength == -1) partialPathLength = path.length;

    final result = SearchResult();
    result.approximate = false;

    Container? currentContainer = this;
    InkObject currentObj = this;

    for (int i = partialPathStart; i < partialPathLength; ++i) {
      final comp = path.getComponent(i);

      if (currentContainer == null) {
        result.approximate = true;
        break;
      }

      final foundObj = currentContainer.contentWithPathComponent(comp);

      if (foundObj == null) {
        result.approximate = true;
        break;
      }

      // Check if next object is a container (for continuing traversal)
      final nextContainer = foundObj is Container ? foundObj : null;
      if (i < partialPathLength - 1 && nextContainer == null) {
        result.approximate = true;
        break;
      }

      currentObj = foundObj;
      currentContainer = nextContainer;
    }

    result.obj = currentObj;
    return result;
  }

  /// Insert content at a specific index.
  void insertContent(InkObject contentObj, int index) {
    content.insert(index, contentObj);

    if (contentObj.parent != null) {
      throw StateError('content is already in ${contentObj.parent}');
    }

    contentObj.parent = this;
    tryAddNamedContent(contentObj);
  }

  /// Add all content from another container.
  void addContentsOfContainer(Container otherContainer) {
    for (final obj in otherContainer.content) {
      obj.parent = this;
      _content.add(obj);
      tryAddNamedContent(obj);
    }
  }

  /// Find content by path component.
  InkObject? contentWithPathComponent(PathComponent component) {
    if (component.isIndex) {
      if (component.index >= 0 && component.index < content.length) {
        return content[component.index];
      }
      return null;
    } else if (component.isParent) {
      return parent;
    } else {
      if (component.name == null) {
        throw StateError('component.name is null');
      }

      final found = namedContent[component.name];
      if (found != null && found is InkObject) {
        return found as InkObject;
      }
      return null;
    }
  }

  /// Build a debug string representation of the hierarchy.
  String buildStringOfHierarchy([
    StringBuffer? sb,
    int indentation = 0,
    InkObject? pointedObj,
  ]) {
    sb ??= StringBuffer();
    const spacesPerIndent = 4;

    void appendIndentation() {
      for (int i = 0; i < spacesPerIndent * indentation; ++i) {
        sb!.write(' ');
      }
    }

    appendIndentation();
    sb.write('[');

    if (hasValidName) {
      sb.write(' ($name)');
    }

    if (this == pointedObj) {
      sb.write('  <---');
    }

    sb.writeln();

    indentation++;

    for (int i = 0; i < content.length; ++i) {
      final obj = content[i];

      if (obj is Container) {
        obj.buildStringOfHierarchy(sb, indentation, pointedObj);
      } else {
        appendIndentation();
        if (obj is StringValue) {
          sb.write('"');
          sb.write(obj.toString().replaceAll('\n', '\\n'));
          sb.write('"');
        } else {
          sb.write(obj.toString());
        }
      }

      if (i != content.length - 1) {
        sb.write(',');
      }

      if (obj is! Container && obj == pointedObj) {
        sb.write('  <---');
      }

      sb.writeln();
    }

    // Named-only content
    final onlyNamed = <String, INamedContent>{};
    for (final entry in namedContent.entries) {
      if (entry.value is InkObject &&
          !content.contains(entry.value as InkObject)) {
        onlyNamed[entry.key] = entry.value;
      }
    }

    if (onlyNamed.isNotEmpty) {
      appendIndentation();
      sb.writeln('-- named: --');

      for (final value in onlyNamed.values) {
        if (value is Container) {
          value.buildStringOfHierarchy(sb, indentation, pointedObj);
          sb.writeln();
        }
      }
    }

    indentation--;

    appendIndentation();
    sb.write(']');

    return sb.toString();
  }

  @override
  String toString() => 'Container(${name ?? "unnamed"})';
}
